File: src/chain-observers/providers/rc-account-provider.ts
import { TAccountName } from "@hiveio/wax";
import { TRegisterEvaluationContext } from "../classifiers/collector-classifier-base";
import { IRcAccount, RcAccountClassifier } from "../classifiers/rc-account-classifier";
import { TProviderEvaluationContext } from "../factories/data-evaluation-context";
import { ProviderBase } from "./provider-base";

export type TRcAccountProvided<TAccounts extends Array<TAccountName>> = {
  [K in TAccounts[number]]: IRcAccount;
};

export interface IRcAccountProviderData<TAccounts extends Array<TAccountName>> {
  rcAccounts: Partial<TRcAccountProvided<TAccounts>>;
};

export interface IRcAccountsProviderOptions {
  accounts: string[];
}

export class RcAccountProvider<
  TAccounts extends Array<TAccountName> = Array<TAccountName>
> extends ProviderBase<IRcAccountsProviderOptions, IRcAccountProviderData<TAccounts>> {
  public readonly rcAccounts = new Set<TAccountName>();

  public pushOptions(options: IRcAccountsProviderOptions): void {
    for(const account of options.accounts)
      this.rcAccounts.add(account);
  }

  public usedContexts(): Array<TRegisterEvaluationContext> {
    const classifiers: Array<TRegisterEvaluationContext> = [];
    for(const rcAccount of this.rcAccounts)
      classifiers.push(RcAccountClassifier.forOptions({ rcAccount }));

    return classifiers;
  }

  public get baseStructure(): IRcAccountProviderData<TAccounts> {
    return {
      rcAccounts: {}
    };
  }

  public async provide(data: TProviderEvaluationContext): Promise<IRcAccountProviderData<TAccounts>> {
    const result = this.baseStructure;

    const rcAccounts = await data.get(RcAccountClassifier);
    for(const rcAccount of this.rcAccounts)
      result.rcAccounts[rcAccount] = rcAccounts.rcAccounts[rcAccount];

    return result as IRcAccountProviderData<TAccounts>;
  }
}

File: python/workerbee/chain_observers/providers/rc_account_provider.py
"""RcAccountProvider — provides RC (resource credit) account data.

Mirrors src/chain-observers/providers/rc-account-provider.ts.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from .._ordered_set import OrderedSet
from ..classifiers.collector_classifier_base import TRegisterEvaluationContext
from ..classifiers.rc_account_classifier import RcAccountClassifier
from .provider_base import ProviderBase

if TYPE_CHECKING:
    from ..factories.data_evaluation_context import DataEvaluationContext
    from ..payloads import RcAccountData, RcAccountsPayload


class RcAccountProvider(ProviderBase):
    def __init__(self) -> None:
        self.accounts: OrderedSet[str] = OrderedSet()

    def push_options(self, options: Any) -> None:
        for account in options["accounts"]:
            self.accounts.add(account)

    def used_contexts(self) -> list[TRegisterEvaluationContext]:
        classifiers: list[TRegisterEvaluationContext] = []
        for rc_account in self.accounts:
            classifiers.append(
                RcAccountClassifier.for_options({"rc_account": rc_account}),
            )
        return classifiers

    async def provide(self, data: DataEvaluationContext) -> RcAccountsPayload:
        rc_data = await data.get(RcAccountClassifier)
        source = rc_data["rc_accounts"]

        rc_accounts: dict[str, RcAccountData | None] = {}
        for rc_account in self.accounts:
            rc_accounts[rc_account] = source.get(rc_account)

        return {"rc_accounts": rc_accounts}
